### PR TITLE
Rudimentary version of Wii IPC determinism.  Ported from my old udpnet branch.

### PR DESCRIPTION
**Pull request by comex**
This version always starts from a blank slate NAND.  Might not be complete.

### CODE REVIEW

**Comment by ZephyrSurfer** [comment]
Can this be rebased?

**Comment by Jofzar** [comment]
@comex Any chance this can be tested and rebased?

**Comment by kkkllleee** [comment]
I am eager to see this come to life.

**Comment by Jofzar** [comment]
I dont like spamming Prs because of the notifcations people get but is there any news on this @comex im really eager to see this :3

**Comment by comex** [comment]
5 months later, blindly rebased.  Haven't tested it yet.

**Comment by comex** [comment]
I'm going to fix the thread safety stuff.

**Comment by Doppler12** [comment]
Gamelist does not appear in linux. 

Hacked a // infront of if (callback(entry)) in FileSearch.cpp simialr to the 648 linux fix. Though it does not break controller config like it did in 648.

Without Save File for Wii game trying to netplay gives a SYSCONF error and fails to start.
Must make a savefile for wii game to netplay and still gives Error for invalid SYSCONF but pressing okay on Dialog box allows netplay to continue, synced.

Done in Linux Mint 17.1
i5 760
GTX 560Ti

**Comment by Jofzar** [comment]
Not sure if you saw my message on irc but there's no problems on windows tried to break it as hard as possible with RNG but nothing desynced

**Comment by JMC47** [comment]
@Doppler12 and I desynced after a while.  When I attempted to close the game I got 3 popups of "Absurdly large aux buffer" and he got a popup of "Can't find Wiimote by Connection Handle 100"

I think our desync probably has to do with the Linux issues, though.

**Comment by Doppler12** [comment]
Alright, here's what I found while testing. If I delete the Wii folder on first start up it gives me a bunch of SYSCONF errors, crashes. Here's a backtrace of the incident
http://pastebin.com/bZJhChej
After that it will boot properly and launch non-netplay emulation with no hitches.

How ever it will give SYSCONF errors on netplay start. Emulation will start after closing the second Error but will segfault after closing the third error box. Thats how me and JMC47 played a few games on netplay, and didn't desync through miracles.
Here's a backtrace for the netplay crash
http://pastebin.com/beiW3yG9

I think there's something wrong with Project-Morations SYSCONF generation in linux. Dolphin-master does not have these issues in any build I've tried.

**Comment by comex** [comment]
@Doppler12 Did you properly install Dolphin to /usr/local/whatever it expects, i.e. do you have icons in the GUI?  I can't reproduce the sysconf issues on my Ubuntu VM when it's installed.

I also can't reproduce gamelist issues.  Try upgrading g++/libstdc++?

**Comment by comex** [comment]
Even though the use of `<regex>` in this PR is not that important, I tend to be sad if I can't use newer features, so I'd prefer if the buildbots could be upgraded.

**Comment by lioncash** [comment]
Seconding comex's request. Being able to use emplace with map data structures would be nice.

**Comment by Jofzar** [comment]
exiting any wii or gc game will crash dolphin causing it to become unresponsive (netplay + normally)

Only started happening on 6046-dirty, not the older rebases.

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
Are the fifo ci bots not updated?  Or is this something else?

**Comment by comex** [comment]
It would be nice if fifoci automatically included crash logs.  But I haven't tried to reproduce myself, so I can do that at some point.

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

**Comment by comex** [comment]
Fixed the FifoCI issue which was masked by a ton of other stupid bugs.  For future reference:

- FifoCI's Dolphin.inis don't have any logging categories enabled (which is apparently off by default - who knew).
- The Verbosity Dolphin.ini setting is only taken into account in DolphinWX.  Otherwise the log level is always set to warning.  Not actually one of the masking bugs, but related to the above one.
- UICommon is horribly named.
- UICommon::CreateDirectories is not called by MainNoGUI, and Android has its own copy of that functionality for some reason.
- SysConf has this whole pointless m_Filename/m_FilenameDefault split.
- When SysConf was not validly loaded (*only* - any setting will work), Dolphin crashes because the wiimote objects were not created, which is dumb.

**Comment by JMC47** [comment]
So is everything fixed now?

**Comment by comex** [comment]
The only thing fixed was a FifoCI issue.  If there are other crashes you'll have to let me know...

**Comment by comex** [comment]
plz review

**Comment by Jofzar** [comment]
I haven't had any regressions at all, no one that I know of has had
regressions/problems on the Australia netplay chat. This is a mostly
windows based group so can't comment on other os's.

Lgtm

**Comment by comex** [comment]
Anyone want to actually review the code?  Pretty please?

File: Source/Core/Common/SysConf.cpp [hunk 1]
```diff
@@ -409,17 +416,17 @@ void SysConf::UpdateLocation()
 	// Clear the old filename and set the default filename to the new user path
 	// So that it can be generated if the file does not exist in the new location
 	m_Filename.clear();
-	m_FilenameDefault =  File::GetUserPath(F_WIISYSCONF_IDX);
+	// Note: We don't use the dummy Wii root here (if in use) because this is
+	// all tied up with the configuration code.  In the future this should
```
Comment (turbedi) [inline Source/Core/Common/SysConf.cpp]: one space too much

File: Source/Core/Core/CoreTiming.cpp [hunk 2]
```diff
@@ -226,6 +226,12 @@ u64 GetIdleTicks()
 void ScheduleEvent_Threadsafe(int cyclesIntoFuture, int event_type, u64 userdata)
 {
 	_assert_msg_(POWERPC, !Core::IsCPUThread(), "ScheduleEvent_Threadsafe from wrong thread");
+	if (Core::g_want_determinism)
+	{
+		ERROR_LOG(POWERPC, "Someone scheduled an off-thread \"%s\" event while netplay or movie play/record "
+		                   "was active.  This is likely to cause a desync.",
```
Comment (turbedi) [inline Source/Core/Core/CoreTiming.cpp]: one space too much

Comment (comex) [inline Source/Core/Core/CoreTiming.cpp]: i prefer two spaces.

**Comment by comex** [comment]
I give up!

**Pull request merged**